### PR TITLE
Support ncm2

### PR DESCRIPTION
**Pull request by roxma**
https://github.com/ncm2/ncm2

Note that the ctx field in the callback may have extra fields in the future, it could be added either by ncm2 core or by ncm2 plugins. It's better to send the whole copy back instead of sending the strongly typed context, which might lose some data.

```rust
       let orig_ctx: Value = serde_json::from_value(rpc::to_value(params.clone())?)?;
       let orig_ctx = &orig_ctx["ctx"];
```

This PR is tested with rls.

### CODE REVIEW

**Comment by roxma** [comment]
Also tested with [ncm2-snipmate](https://github.com/ncm2/ncm2-snipmate) and [ncm2-ultisnips](https://github.com/ncm2/ncm2-ultisnips)

```vim
    " assuming your using vim-plug: https://github.com/junegunn/vim-plug
    Plug 'ncm2/ncm2'
    " ncm2 requires nvim-yarp
    Plug 'roxma/nvim-yarp'

    autocmd BufEnter * call ncm2#enable_for_buffer()
    set completeopt=noinsert,menuone,noselect
    set shortmess+=c

    au TextChangedI * call ncm2#auto_trigger()

    inoremap <c-c> <ESC>

    inoremap <expr> <CR> (pumvisible() ? "\<c-y>\<cr>" : "\<CR>")
    inoremap <expr> <Tab> pumvisible() ? "\<C-n>" : "\<Tab>"
    inoremap <expr> <S-Tab> pumvisible() ? "\<C-p>" : "\<S-Tab>"

    Plug 'ncm2/ncm2-bufword'

    Plug 'autozimu/LanguageClient-neovim'
    let g:LanguageClient_serverCommands = {
    \ 'go': ['go-langserver'],
    \ 'rust': ['rls'],
    \ }
    let g:LanguageClient_loggingFile = '/tmp/lc.log'
    let g:LanguageClient_loggingLevel = 'DEBUG'

    " a json file with settings for go-langserver:
    " {
    " 	"initializationOptions": {
    " 		"gocodeCompletionEnabled": true,
    " 		"funcSnippetEnabled": true
    " 	}
    " }
    let g:LanguageClient_settingsPath = $WORKSPACE_DIR . '/.vim/settings.json'

    Plug 'ncm2/ncm2-ultisnips'
    Plug 'SirVer/ultisnips'

    inoremap <silent> <expr> <CR> ncm2_ultisnips#expand_or("\<CR>", 'n')

    " c-j c-k for moving in snippet
    imap <c-u> <Plug>(ultisnips_expand)
    let g:UltiSnipsExpandTrigger		= "<Plug>(ultisnips_expand)"
    let g:UltiSnipsJumpForwardTrigger	= "<c-j>"
    let g:UltiSnipsJumpBackwardTrigger	= "<c-k>"
    let g:UltiSnipsRemoveSelectModeMappings = 0
```

File: src/languageclient.rs [hunk 1]
```diff
@@ -548,6 +548,47 @@ impl State {
...
+            json!([{
+                "name": format!("LanguageClient_{}", languageId),
+                "priority": 9,
+                "scope": [languageId],
+                "complete_pattern": trigger_patterns,
+                "mark": "LC",
+                "on_complete": REQUEST__NCM2OnComplete,
+            }]),
+        )?;
+        info!("End register NCM source");
```
Comment (autozimu) [inline src/languageclient.rs]: Please update the message content.

File: src/types.rs [hunk 2]
```diff
@@ -430,6 +431,24 @@ pub struct NCMRefreshParams {
...
+    pub filetype: String,
+    pub typed: String,
+    pub filepath: String,
+    pub scope: String,
+    pub startccol: u64,
+    pub base: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct NCM2OnCompleteParams {
```
Comment (autozimu) [inline src/types.rs]: I don't think this is ever used.

**Review by autozimu — CHANGES_REQUESTED** [review]

**Comment by autozimu** [comment]
Thanks.

**Comment by run27017** [comment]
What effects? How about auto completeion?

**Comment by martskins** [comment]
@run27017 I don't think I understand the question. Care to elaborate?

**Comment by run27017** [comment]
@martskins Sorry about it. I am seeking the common integration method of ncm2 and LanguageClient-neovim, which supporting the auto completion. It seems this PR is not related to it, and I will search the data in other place.

**Pull request merged**